model/src/pyrenew/metaclasses.py

```diff
@@ -0,0 +1,251 @@
+#!/usr/bin/env/python
+# -*- coding: utf-8 -*-
+
+"""
+Observation helper classes
+"""
+
+from abc import ABCMeta, abstractmethod
+
+import jax
+import polars as pl
+from numpyro.infer import MCMC, NUTS
+from pyrenew.mcmcutils import spread_draws
+
+
+def _assert_sample_and_rtype(
+    rp: "RandomProcess", skip_if_none: bool = True
+) -> None:
+    """Return type-checking for RandomProcess's sample function
+
+    Objects passed as `RandomProcess` should (a) have a `sample()` method that
+    (b) returns either a tuple or a named tuple.
+
+    Parameters
+    ----------
+    rp : RandomProcess
+        Random process to check.
+    skip_if_none: bool
+        When `True` it returns if `rp` is None.
+
+    Returns
+    -------
+    None
+    """
+
+    # Addressing the None case
+    if (rp is None) and (not skip_if_none):
+        Exception(
+            "The passed object cannot be None. It should be RandomProcess"
+        )
+    elif skip_if_none and (rp is None):
+        return None
+
+    if not isinstance(rp, RandomProcess):
+        raise Exception(f"{rp} is not an instance of RandomProcess.")
+
+    # Otherwise, checking for the sample function (must have one)
+    # with a defined rtype.
+    try:
+        sfun = rp.sample
+    except Exception:
+        raise Exception(
+            f"The RandomProcess {rp} does not have a sample function."
+        )  # noqa: E722
+
+    # Getting the return annotation (if any)
+    rettype = sfun.__annotations__.get("return", None)
+
+    if rettype is None:
+        raise Exception(
+            f"The RandomProcess {rp} does not have return type "
+            + "annotation."
+        )
+
+    try:
+        if not isinstance(rettype(), tuple):
+            raise Exception(
+                f"The RandomProcess {rp}'s return type annotation is not"
+                + "a tuple"
+            )
+    except Exception:
+        raise Exception(
+            f"There was a problem when trying to initialize {rettype}."
+            + "the rtype of the random process should be a tuple or a namedtuple"
+            + " with default values."
+        )
+
+    return None
+
+
+class RandomProcess(metaclass=ABCMeta):
```

Comment:
I think given the new architecture we may want to rename this simply `RandomVariable`; we also need to rewrite the docstring given how we're now using it.